src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,382 @@
+        if is_category_dtype(dtype) and elements is None:
+            raise InvalidArgument(
+                'Must provide an elements strategy for category dtypes'
+            )
+
+    def __repr__(self):
+        return 'column(%r, dtype=%r, elements=%r)' % (
+            self.name, self.dtype, self.elements,
+        )
+
+
+def columns(names_or_number, dtype=None, elements=None):
+    st.check_type(
+        (Iterable,) + integer_types, names_or_number, 'names_or_number'
+    )
+    try:
+        names = list(names_or_number)
+    except TypeError:
+        names = [None] * names_or_number
+    return [
+        column(name=n, dtype=dtype, elements=elements) for n in names
+    ]
+
+
+@st.composite
+def data_frames(
+    draw,
+    rows_or_columns=None,
+    columns=None, rows=None, index=None, min_size=0, max_size=None
+):
+    """Provides a strategy for producing a pandas.DataFrame.
+
+    Arguments:
+        rows_or_columns: A special argument designed for convenient use as a
+        positional argument. It will be passed as one of the rows or columns
+        arguments according to the following rules:
+
+        * An error is raised if all three of rows_or_columns, rows and columns
+          are not None.
+        * If rows_or_columns is not None and one of rows or columns is not None
+          then rows_or_columns will be passed as the other.
+        * If rows_or_columns is not None and rows and columns are both None
+          then the type of the value passed will be used. If it is a strategy
+          then it will be passed as rows, else it will be passed as columns.
+
+        columns: An iterable of column objects describing the shape of the
+            generated DataFrame.
+
+        rows: A strategy for generating a row object. Should generate anything
+            that could be passed to pandas as a list - e.g. dicts, tuples.
+
+            At least one of rows and columns must be provided. If both are
+            provided then the generated rows will be validated against the
+            columns and an error will be raised if they don't match.
+
+            Caveats on using rows:
+
+            * In general you should prefer using columns to rows, and only use
+              rows if the columns interface is insufficiently flexible to
+              describe what you need - you will get better performance and
+              example quality that way.
+            * If you provide rows and not columns, then the shape and dtype of
+              the resulting DataFrame may vary. e.g. if you have a mix of int
+              and float in the values for one column in your row entries, the
+              resulting DataFrame will sometimes have an integral dtype and
+              sometimes a float.
+
+        index: a sequence or a strategy for generating a sequence. It will
+            be used as the index for the resulting series. When it is longer
+            than the result it will be truncated to the right side.
```

Comment:
I think on reflection what I'm actually going to do is make the index argument an exact index.